### ISSUE TITLE
Allow exporting self-contained SVG files

### ISSUE THREAD

**Issue by raichu** [fabricjs/fabric.js#519]
Currently, the image elements are exported with their "ordinary" URLs. Since SVG supports data URIs, it should be possible to embed these images and make the SVG output of the canvas self-contained.
##

**Comment by tiptronic** [comment]
Is there any interest in adding this option today?

**Comment by asturur** [comment]
I remember we did something for it. Did you give a look at image toSVG implementation?

**Comment by tiptronic** [comment]
Yes - I saw the modifications, but these seem only to affect the situation, if I want to get out a single layer/object.

In my case, I have a canvas contiaining some (layered) objects. E.g. Rect, Image (img.src="..") and an Oval.
When exporting to SVG using `canvas.toSVG` I want to embed the image as base64. So what I do is:
`
canvas.toSVG({suppressPreamble:true}, function(svg, a, b) {
     return svg;
}
`
But here I get only the full svg-element containing the `xlink:href` as `src`. 
Could it be, that I MUST overwrite the `getSvgSrc`prototype to get the dataUrl or an image?

**Comment by tiptronic** [comment]
It seems this is everything required:

```javascript
fabric.Image.prototype.getSvgSrc = function() {
       return this.toDataURL();
}
canvas.toSVG({ suppressPreamble: true });
```

**Comment by asturur** [comment]
Well i remember we plan for this to be an option.
But to do so, we need to convert the image at export time. Not a big deal, since if this is what the dev need, at some point he has to do it too.

I have not time to prioritize this over the other open things.

You can override the method as you pointed out, using the toDataUrl without any transformation  ( or if the object is rotated, you will get a rotated dataUrl that is what you do NOT want )

**Comment by tiptronic** [comment]
OK - understood. Thanks for the explanation - so to keep the original image I have to convert the originalElement toDataURL if I intend to get the 'raw' (unmodified) image, right?

**Comment by asturur** [comment]
the Object.toDataUrl offers you the ability to skip transformations.

http://fabricjs.com/docs/fabric.Object.html#toDataURL

The withoutTransform option.

**Comment by tiptronic** [comment]
@asturur Bummer! I didn't see that in the docs - sorry! That's perfect! Grazie.
